crates/pixi_build_frontend/src/protocols/mod.rs

```diff
@@ -1,4 +1,381 @@
+use thiserror::Error;
+use tokio::{
+    io::{AsyncBufReadExt, BufReader, Lines},
+    process::ChildStderr,
+    sync::{oneshot, Mutex},
+};
+
+use crate::{
+    jsonrpc::{stdio_transport, Receiver, RpcParams, Sender},
+    tool::Tool,
+    CondaBuildReporter, CondaMetadataReporter,
+};
+
 pub(super) mod conda_build;
 mod error;
 pub(super) mod pixi;
+pub(super) mod rattler_build;
+pub(super) mod stderr;
+
+#[derive(Debug, Error, Diagnostic)]
+pub enum InitializeError {
+    #[error("failed to setup communication with the build backend, an unexpected io error occurred while communicating with the pixi build backend")]
+    #[diagnostic(help("Ensure that the project manifest contains a valid [build] section."))]
+    Io(#[from] std::io::Error),
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    Protocol(#[from] ProtocolError),
+}
+
+#[derive(Debug, Error, Diagnostic)]
+pub enum ProtocolError {
+    #[error("failed to communicate with the build backend ({0})")]
+    #[diagnostic(help(
+        "Ensure that the build backend implements the JSON-RPC protocol correctly."
+    ))]
+    JsonRpc(String, #[source] ClientError),
+    #[error("received invalid response from the build backend ({0}) when calling '{1}'")]
+    ParseError(String, String, #[source] serde_json::Error),
+    #[error(transparent)]
+    #[diagnostic(
+        transparent,
+        help("This error originates from the build backend specified in the project manifest.")
+    )]
+    BackendError(
+        #[from]
+        #[diagnostic_source]
+        BackendError,
+    ),
+    #[error("the build backend ({0}) does not implement the method '{1}'")]
+    #[diagnostic(help(
+        "This is often caused by the build backend incorrectly reporting certain capabilities. Consider contacting the build backend maintainers for a fix."
+    ))]
+    MethodNotImplemented(String, String),
+
+    #[error("pipe of stderr stopped earlier than expected")]
+    StdErrPipeStopped,
+}
+
+impl ProtocolError {
+    pub fn from_client_error(backend_identifier: String, err: ClientError, method: &str) -> Self {
+        match err {
+            Error::Call(err) if err.code() > -32001 => Self::BackendError(BackendError::from(err)),
+            Error::Call(err) if err.code() == ErrorCode::MethodNotFound.code() => {
+                Self::MethodNotImplemented(backend_identifier, method.to_string())
+            }
+            Error::ParseError(err) => Self::ParseError(backend_identifier, method.to_string(), err),
+            e => Self::JsonRpc(backend_identifier, e),
+        }
+    }
+}
+
+/// Protocol trait that is responsible to setup and communicate with the backend.
+/// This allow us to hide the jsonrpc communication hidden in this protocol.
+/// This protocol is generic over the manifest what are passed to the build backends.
+/// This means that, for rattler-build, the manifest is a recipe.yaml file,
+/// and for pixi it's a pixi.toml/pyproject.toml file.
+#[async_trait::async_trait]
```

Comment:
We could try using rust async fn in trait support?

Reply:
Hmm actually, I rather that this a struct and named something like `JsonRPCBuildProtocol`. Essentially, pixi-build-backend and rattler-build-backend are user of this protocol and these things:

```rust
    /// Returns a unique identifier for the backend.
    fn backend_identifier(&self) -> &str;

    /// Returns the client used to communicate with the backend.
    fn client(&self) -> &Client;

    /// Returns a unique identifier for the build.
    fn build_id(&self) -> usize;

    /// Returns the handle to the stderr of the backend process.
    fn stderr(&self) -> Option<Arc<Mutex<Lines<BufReader<ChildStderr>>>>>;
```

with the exclusion of `stderr` can be passed in. This feels like it could all just be passed into the struct I've mentioned above. It feels like this is emulating inheritance where it need not be. `stderr` might be a method that can be implemented on the struct directly.

Basically, its all the same protocol but just using different data. A trait, in my opinion would make more sense, if we would abstract over different kind of protocols, like XMLRPC vs JsonRPC.